### PR TITLE
InMemoryLogger has get_timeseries() method

### PR DESCRIPTION
**Pull request by growlix**
Solution to #513

### CODE REVIEW

File: composer/loggers/in_memory_logger.py [hunk 1]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
             self.data[k].append((timestamp, log_level, v))
         self.most_recent_values.update(data)
         self.most_recent_timestamps.update({k: timestamp for k in data})
+
+    def get_timeseries(self, metric: str, time_unit: Union[str, None] = None):
```
Comment (hanlint) [inline composer/loggers/in_memory_logger.py:79]: Would be good to add a quick test into pytest on this method, e.g.:
```
logger = InMemoryLogger(..)

# patch the data directly
logger.data = { ..}

assert logger.get_timeseries(...) ==
```




```

Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:79]: For TimeUnit, we generally take the string or enum value:

```suggestion
    def get_timeseries(self, metric: str, time_unit: Optional[Union[str, TimeUnit]] = None):
```

File: composer/loggers/in_memory_logger.py [hunk 2]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+        """Returns logged data as dict containing values of a desired metric over time.
+
+        Args:
+            metric (str): Metric of interest. Must be present in self.data.keys().
+            time_unit(str or None, optional) Which unit of time to return. Options =
```
Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:84]:
```suggestion
            time_unit (str or None, optional) Which unit of time to return. Options =
```

File: composer/loggers/in_memory_logger.py [hunk 3]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+                values of that time unit for the corresponding index of the metric. For
+                example, plotting timeseries['accuracy/val'] vs. timeseries['epoch'] would
+                plot validation accuracy as a function of epoch. And if
+                timeseries['accuracy/val'][5] == 41.2, and timeseries['batch'][5] == 680,
+                that means that 'accuracy/val' was equal to 41.2 on the 680th batch.
+        """
+
+        # Check time_unit string
+        valid_time_strings = {"batch", "epoch", "sample", "token", "duration", None}
+        if time_unit not in valid_time_strings:
```
Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:101]: you can try casting via `time_unit = TimeUnit(time_unit)` -- if it's an illegal unit, then a ValueError will be raised. So no need to validate manually,

File: composer/loggers/in_memory_logger.py [hunk 4]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+        Args:
+            metric (str): Metric of interest. Must be present in self.data.keys().
+            time_unit(str or None, optional) Which unit of time to return. Options =
+                {"batch", "epoch", "sample", "token", "duration", None}. If ``None``,
+                will return all available units of time. Default=``None``.
+
+        Returns:
+            timeseries (dict): Dictionary in which one key is ``metric``, and the
+                associated value is a list of values of that metric. The remaining key(s)
+                are each a unit of time, and the associated values are each a list of
```
Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:91]: From a typing perspective, I'm not a huge fan of mixing the value types of a Dict.The annotation of the returned dictionary would be something like `Dict[Union[str, TimeUnit], Union[List[TLogData], List[Time]]]`.

I'm wondering whether it would make sense to break this into returning a tuple where one element is the data and the other is a dictionary of the timestamps -- e.g. `Tuple[List[TLogData], Dict[TimeUnit, List[Time]]`?

File: composer/loggers/in_memory_logger.py [hunk 5]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+                values of that time unit for the corresponding index of the metric. For
+                example, plotting timeseries['accuracy/val'] vs. timeseries['epoch'] would
+                plot validation accuracy as a function of epoch. And if
+                timeseries['accuracy/val'][5] == 41.2, and timeseries['batch'][5] == 680,
+                that means that 'accuracy/val' was equal to 41.2 on the 680th batch.
+        """
+
+        # Check time_unit string
+        valid_time_strings = {"batch", "epoch", "sample", "token", "duration", None}
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/loggers/in_memory_logger.py:100]: Validation should not need to be done manually -- see comment below.

File: composer/loggers/in_memory_logger.py [hunk 6]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+    def get_timeseries(self, metric: str, time_unit: Union[str, None] = None):
+        """Returns logged data as dict containing values of a desired metric over time.
+
+        Args:
+            metric (str): Metric of interest. Must be present in self.data.keys().
+            time_unit(str or None, optional) Which unit of time to return. Options =
+                {"batch", "epoch", "sample", "token", "duration", None}. If ``None``,
+                will return all available units of time. Default=``None``.
+
+        Returns:
```
Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:88]: Can you add a doctest? I presume this helper is for matplotlib, in which case it would be really nice to show how the results from here can be passed directly into a pyplot.

File: composer/loggers/in_memory_logger.py [hunk 7]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
         self.most_recent_values.update(data)
         self.most_recent_timestamps.update({k: timestamp for k in data})
+
+    def get_timeseries(self, metric: str, time_unit: Union[str, None] = None):
+        """Returns logged data as dict containing values of a desired metric over time.
+
+        Args:
+            metric (str): Metric of interest. Must be present in self.data.keys().
+            time_unit(str or None, optional) Which unit of time to return. Options =
+                {"batch", "epoch", "sample", "token", "duration", None}. If ``None``,
```
Comment (ravi-mosaicml) [inline composer/loggers/in_memory_logger.py:85]: I don't think it's necessary to enumerate the units here. Also thoughts about removing the unit parameter and instead always returning all units?

File: composer/loggers/in_memory_logger.py (hunk 3 above)
Comment (hanlint) [reply composer/loggers/in_memory_logger.py:101]: `TimeUnit` allows for `duration`, but `TimeStamp` does not have the `duration` field.

File: composer/loggers/in_memory_logger.py (hunk 4 above)
Comment (hanlint) [reply composer/loggers/in_memory_logger.py:91]: The return type here should always be `Dict[str, TLogData]`. Even for time units, the code below does:
```
timeseries[time_unit].append(getattr(timestamp, time_unit).value)
```

where `time_unit` is a string, and the latter is compatible with `TLogData`.

File: composer/loggers/in_memory_logger.py [hunk 8]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+                will return all available units of time. Default=``None``.
+
+        Returns:
+            timeseries (dict): Dictionary in which one key is ``metric``, and the
+                associated value is a list of values of that metric. The remaining key(s)
+                are each a unit of time, and the associated values are each a list of
+                values of that time unit for the corresponding index of the metric. For
+                example, plotting timeseries['accuracy/val'] vs. timeseries['epoch'] would
```
[the remark this answers is not in the document]
Comment (growlix) [reply composer/loggers/in_memory_logger.py:93]: It does not support returning multiple metrics. Different metrics may have different sampling resolutions and hence different length time series. I think this would require returning nested dicts, which feels like worse UX to me than making multiple calls to `.get_timeseries()`, but others may disagree and also it's possible there's a nice solution I'm not thinking of.

[the remark this answers is not in the document]
Comment (hanlint) [reply composer/loggers/in_memory_logger.py:93]: Agreed!

File: composer/loggers/in_memory_logger.py (hunk 1 above)
Comment (growlix) [reply composer/loggers/in_memory_logger.py:79]: TimeUnits are a little clunky here because TimeStamp fields don't follow the same convention: they're named `epoch`, `batch`, `batch_in_epoch`, etc.

File: composer/loggers/in_memory_logger.py [hunk 9]
```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+            timeseries[metric].append(metric_value)
+            timestamp = datapoint[0]
+            if time_unit:
+                timeseries[time_unit].append(getattr(timestamp, time_unit).value)
+            else:
+                # Iterate through time units and add them all!
+                for field in timestamp._fields:
+                    time_value = getattr(timestamp, field).value
+                    timeseries[field].append(time_value)
+        return timeseries
```
Comment (growlix) [inline composer/loggers/in_memory_logger.py:124]: Having type check problems here: `Expression of type "defaultdict[Unknown, list[Unknown]]" cannot be assigned to return type "Dict[str, TLogData]"`. Suggested solutions @ravi-mosaicml, @hanlint?

File: composer/loggers/in_memory_logger.py (hunk 3 above)
Comment (ravi-mosaicml) [reply composer/loggers/in_memory_logger.py:101]: If it's just one value that's not supported, can we manually check for that and raise an error? Not a fan of manually repeating this here.

File: composer/loggers/in_memory_logger.py (hunk 9 above)
Comment (ravi-mosaicml) [reply composer/loggers/in_memory_logger.py:124]: Easiest way to fix would be to not use a defaultdict and instead use a dict

**Comment by growlix** [comment]
I removed the `time_unit` argument and return time series for all time units.

Mostly because`TimeStamp` fields can't be accessed using `TimeUnits` because they don't follow the same convention. `TimeStamp` fields are named e.g. `"epoch"` and `"batch"`, and there exist TimeStamp fields that aren't TimeUnits (e.g. `"batch_in_epoch"`). This makes me wonder if we should create a `TimeUnit` -> `TimeStamp` field mapping somewhere for easy conversion.

Anyway, this method is a lot simpler now.

**Comment by growlix** [comment]
Some data in the logger are stored as tensors, such that calling `in_mem_logger.get_timeseries("accuracy/val")` returns `[tensor(0.2287, devic...='cuda:0'), tensor(0.4015, devic...='cuda:0'), ...]`. Should we check for whether the data are tensors, and if so convert them to lists?

**Comment by hanlint** [comment]
I think returning them as numpy arrays would be reasonable.

File: composer/loggers/in_memory_logger.py [hunk 10]
```diff
@@ -74,3 +77,58 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
...
+        Example:
+            .. testcode::
+
+                import matplotlib.pyplot as plt
+
+                from composer.core.logging import LogLevel
+                from from composer.loggers import InMemoryLogger
+
+                in_mem_logger = InMemoryLogger(LogLevel.BATCH)
+                in_mem_logger = populate_in_memory_logger(in_mem_logger, datafield="accuracy/val")
```
Comment (hanlint) [inline composer/loggers/in_memory_logger.py:106]: Might be better to directly populate it in the testcode here with some minimal data, e.g.
```
# for this example, directly populate the data
in_mem_logger.data = {'accuracy/val': [(1, blah, 0.4), (2, blah, 1.0)]}
timeseries = in_mem_logger.get_timeseries("accuracy/val")
timeseries
>>> [0.4, 1.0]
```

Comment (growlix) [reply composer/loggers/in_memory_logger.py]: I populate in the example now, but it's pretty verbose (as you can see) because it requires `Timestamp`s. I could go either way!

File: tests/test_logger.py [hunk 11]
```diff
@@ -149,3 +151,28 @@ def test_in_memory_logger(dummy_state: State):
     # the most recent values should have just the last call to epoch
     assert in_memory_logger.most_recent_values["epoch"] == "should_be_recorded_and_override"
     assert in_memory_logger.most_recent_timestamps["epoch"].batch == 1
+
+    # test get_timeseries()
```
Comment (hanlint) [inline tests/test_logger.py:155]: this should be a separate test

File: docs/source/doctest_fixtures.py [hunk 12]
```diff
@@ -110,3 +114,27 @@ def Trainer(fake_ellipses='...', *args, **kwargs):
 
 # patch composer so that 'from composer import Trainer' calls do not override change above
 composer.Trainer = Trainer
+
+
+def populate_in_memory_logger(logger: InMemoryLogger,
```
Comment (hanlint) [inline docs/source/doctest_fixtures.py:119]: this can be removed

**Review by hanlint — APPROVED** [review]
LGTM, two minor comments, and then good to merge.
